http:support for adding dynamic request headers to the upstream requests

khgandhi opened the pull request:
created a new PR as a continuation to PR #1131 which could not be reopened after it was closed. 

@mattklein123 @htuch @RomanDzhabarov 

GT #1016

mattklein123 commented on include/envoy/http/access_log.h:
```diff
@@ -114,6 +114,16 @@ class RequestInfo {
+  /**
+   * Set the downstream address
+   */
+  virtual void setDownstreamAddress(const std::string& address) PURE;
```
Please remove this from the interface per the previous discussion. It doesn't make sense to be able to set the downstream address on the request info at the interface level IMO.

khgandhi replied:
@RomanDzhabarov i think @mattklein123  is suggesting to move it out of connection_manager_impl itself (it was already present) and just keep it in the request_info because the connection_manager_impl has a reference to request_info as well.

mattklein123 commented on source/common/http/access_log/request_info_impl.h:
```diff
@@ -54,6 +61,7 @@ struct RequestInfoImpl : public RequestInfo {
   uint64_t response_flags_{};
   Upstream::HostDescriptionConstSharedPtr upstream_host_{};
   bool hc_request_{};
+  std::string downstream_address_;
```
Can we avoid keeping a copy of downstream_address_ in both here as well as the connection manager. I would probably just have it here, delete from connection manager, then allow setting via a public `setDownstreamAddress` which is not actually part of the interface but is public to the connection manager.

khgandhi replied:
@mattklein123 the downstream_address_ is required on the request_info object because it gets accessed in the req_header_formatter.cc (line 79 [here](https://github.com/khgandhi/envoy/blob/43ccc9accf4aa6fd663914d84c5a3846d775bfe5/source/common/router/req_header_formatter.cc#L79)). I am not sure how would the formatter access the downstream_address_ if is set on the connection_manager.

mattklein123 replied:
I know it's required on the request info. I'm saying to not duplicate it in the connection manager itself. Use the one in the request info. You are copying the same string twice.

khgandhi replied:
@mattklein123 ok.. i will address this.

mattklein123 commented on source/common/router/req_header_formatter.h:
```diff
@@ -0,0 +1,100 @@
+  virtual std::string format(const Envoy::Http::AccessLog::RequestInfo& request_info) const PURE;
+};
+
+typedef std::shared_ptr<HeaderFormatter> HeaderFormatterSharedPtr;
```
Why does this need to be a shared_ptr? I don't think it does.

khgandhi replied:
@RomanDzhabarov I tried moving this to a unique_ptr and also tried using std::move inside the parse method in RequestHeaderPrarser but it still looks like the compiler is not happy with it. Once i change it to a shared_ptr it starts working.

mattklein123 commented on source/common/router/req_header_formatter.h:
```diff
@@ -0,0 +1,100 @@
+class HeaderFormatter {
+
+public:
+  virtual ~HeaderFormatter(){};
```
nit: remove trailing ';' here and everywhere else this happens.

khgandhi replied:
will address this.

mattklein123 commented on source/common/router/req_header_formatter.h:
```diff
@@ -0,0 +1,100 @@
+class PlainHeaderFormatter : public HeaderFormatter {
+
+public:
+  virtual ~PlainHeaderFormatter(){};
```
del not needed.

mattklein123 commented on source/common/router/req_header_formatter.h:
```diff
@@ -0,0 +1,100 @@
+  };
+
+private:
+  std::string static_value_;
```
const

khgandhi replied:
@RomanDzhabarov i am not sure why the static_value_ should be defined a const. it is getting assigned a value in the constructor PlainHeaderFormatter.

mattklein123 replied:
make const, and then initialize in the constructor initializer list. Please do this anywhere else applicable.

mattklein123 commented on source/common/router/router.cc:
```diff
@@ -159,6 +159,7 @@ Http::FilterHeadersStatus Filter::decodeHeaders(Http::HeaderMap& headers, bool e
 
   // Determine if there is a route entry or a redirect for the request.
   route_ = callbacks_->route();
+
```
Undo whitespace modifications in this file.

mattklein123 commented on source/common/router/router.cc:
```diff
@@ -228,8 +229,8 @@ Http::FilterHeadersStatus Filter::decodeHeaders(Http::HeaderMap& headers, bool e
   }
-
-  route_entry_->finalizeRequestHeaders(headers);
+  log().debug("downstream address: '{}'\n", callbacks_->downstreamAddress());
```
del, not needed.

mattklein123 commented on source/common/router/req_header_formatter.h:
```diff
@@ -0,0 +1,100 @@
+ * a formatter that expands the request header variable to a value based on info in RequestInfo.
+ */
+class RequestHeaderFormatter : public HeaderFormatter, Logger::Loggable<Logger::Id::config> {
+
```
del newline. (In general make code look like other code)

mattklein123 commented on source/common/router/req_header_formatter.h:
```diff
@@ -0,0 +1,100 @@
+class RequestHeaderFormatter : public HeaderFormatter, Logger::Loggable<Logger::Id::config> {
+
+public:
+  virtual ~RequestHeaderFormatter(){};
```
del

mattklein123 commented on source/common/router/req_header_formatter.h:
```diff
@@ -0,0 +1,100 @@
+// forward declaring the typedef and class
+class RequestHeaderParser;
+
+typedef std::shared_ptr<RequestHeaderParser> RequestHeaderParserSharedPtr;
```
Why does this need to be a shared_ptr? I don't think it does.

khgandhi replied:
without that it does not let me copy the object over to objects like RouteEntryImplBase where it is actually invoked to start the actual evaluation of the headers https://github.com/khgandhi/envoy/blob/43ccc9accf4aa6fd663914d84c5a3846d775bfe5/source/common/router/config_impl.cc#L87

mattklein123 replied:
Use std::move and move semantics. @khgandhi please reach out to 1:1 to @RomanDzhabarov for C++ questions/help so we can move this along. He can help you.

khgandhi replied:
Thanks @mattklein123 .

@RomanDzhabarov.. would it make sense to not use a ptr at all here ?

RomanDzhabarov replied:
You can use unique_ptr and return that from the parse method. And use std::move everywhere. Does not seem that shared is needed here.

mattklein123 commented on source/common/router/req_header_formatter.h:
```diff
@@ -0,0 +1,100 @@
+  std::string static_value_;
+};
+
+// forward declaring the typedef and class
```
del, obvious. del newline at 66

RomanDzhabarov commented on source/common/router/req_header_formatter.h:
```diff
@@ -0,0 +1,100 @@
+      Http::HeaderMap& headers, const Http::AccessLog::RequestInfo& requestInfo,
+      const std::list<std::pair<Http::LowerCaseString, std::string>>& requestHeadersToAdd) const;
+
+  std::unordered_map<Http::LowerCaseString, HeaderFormatterSharedPtr, Http::LowerCaseStringHasher>
```
return type for th headerFormatterMap should be a reference not the copy.

khgandhi commented:
@mattklein123 @RomanDzhabarov 
updated the PR with the docs and incorporated the feedback. Please review.

mattklein123 commented on include/envoy/http/access_log.h:
```diff
@@ -114,6 +114,16 @@ class RequestInfo {
+  /**
+   * Set the downstream address.
+   */
+  virtual void setDownstreamAddress(const std::string& address) PURE;
```
I've already mentioned this several times, but this setter should not be on the public interface. It doesn't make sense.

RomanDzhabarov replied:
@khgandhi you can remove this from the interface and directly set downstream_address_ (as it's a public member of a struct). Similar to bytesReceived: https://github.com/lyft/envoy/blob/master/source/common/http/access_log/request_info_impl.h#L51
https://github.com/lyft/envoy/blob/master/source/common/http/access_log/request_info_impl.h#L19

khgandhi commented:
@mattklein123 @RomanDzhabarov  incorporated the feedback.

RomanDzhabarov commented:
will look today, great thanks!

RomanDzhabarov commented on include/envoy/http/access_log.h:
```diff
@@ -114,6 +114,11 @@ class RequestInfo {
   virtual void healthCheck(bool is_hc) PURE;
+
+  /**
+   * Get the down stream address.
```
nit: downstream

RomanDzhabarov commented on source/common/http/access_log/request_info_impl.h:
```diff
@@ -46,6 +46,11 @@ struct RequestInfoImpl : public RequestInfo {
   void healthCheck(bool is_hc) override { hc_request_ = is_hc; }
 
+  /**
+   * Get the down stream address.
```
ditto

RomanDzhabarov commented on source/common/router/req_header_formatter.h:
```diff
@@ -0,0 +1,88 @@
+  format(const Envoy::Http::AccessLog::RequestInfo& request_info) const PURE;
+};
+
+typedef std::shared_ptr<HeaderFormatter> HeaderFormatterPtr;
```
what was the issue when you used unique_ptr here?

khgandhi replied:
When i change this to the unique_ptr, i add move semantics in req_header_formatter.cc as follows:

RequestHeaderParser RequestHeaderParser::parse(const Json::Object& config) {
  RequestHeaderParser request_header_parser;
  std::unordered_map<Http::LowerCaseString, HeaderFormatterPtr, Http::LowerCaseStringHasher>
      header_formatter_map;
  if (config.hasObject("request_headers_to_add")) {
    std::vector<Json::ObjectSharedPtr> request_headers =
        config.getObjectArray("request_headers_to_add");
    for (const Json::ObjectSharedPtr& header : request_headers) {
      if (!header->getString("key").empty() && !header->getString("value").empty()) {
        ENVOY_LOG(debug, "adding key {} to header formatter map", header->getString("key"));
        HeaderFormatterPtr header_formatter =
            RequestHeaderParser::parseInternal(header->getString("value"));
        request_header_parser.header_formatter_map_.emplace(
            Http::LowerCaseString(header->getString("key")), **std::move(header_formatter)**);
      }
    }
  }

  return request_header_parser;

However, i see the following compilation error
error: call to implicitly-deleted copy constructor of 'std::pair<const Envoy::Http::LowerCaseString, std::unique_ptr<Envoy::Router::HeaderFormatter, std::default_delete<Envoy::Router::HeaderFormatter> > >'

RomanDzhabarov replied:
can you try, map_.insert(std::make_pair(Http::LowerCaseString(header->getString("key")), std::move(header_formatter))); ?

khgandhi replied:
@RomanDzhabarov That did not work either. :(. I got the same error.

RequestHeaderParser RequestHeaderParser::parse(const Json::Object& config) {
 36   RequestHeaderParser request_header_parser;
 37   std::unordered_map<Http::LowerCaseString, HeaderFormatterPtr, Http::LowerCaseStringHasher>
 38       header_formatter_map;
 39   if (config.hasObject("request_headers_to_add")) {
 40     std::vector<Json::ObjectSharedPtr> request_headers =
 41         config.getObjectArray("request_headers_to_add");
 42     for (const Json::ObjectSharedPtr& header : request_headers) {
 43       if (!header->getString("key").empty() && !header->getString("value").empty()) {
 44         ENVOY_LOG(debug, "adding key {} to header formatter map", header->getString("key"));
 45         HeaderFormatterPtr header_formatter =
 46             RequestHeaderParser::parseInternal(header->getString("value"));
 47         request_header_parser.header_formatter_map_.insert(std::make_pair(Http::LowerCaseString(header->getString("key")),std::move(header_formatter)));
 48       }
 49     }
 50   }
 51
 52   return request_header_parser;


error:

/usr/bin/../lib/gcc/x86_64-linux-gnu/5.4.0/../../../../include/c++/5.4.0/ext/new_allocator.h:120:23: error: call to implicitly-deleted copy constructor of 'std::pair<const Envoy::Http::LowerCaseString, std::unique_ptr<Envoy::Router::HeaderFormatter, std::default_delete<Envoy::Router::HeaderFormatter> > >'
        { ::new((void *)__p) _Up(std::forward<_Args>(__args)...); }

RomanDzhabarov replied:
I'll look into it

RomanDzhabarov replied:
ok, see below, this is fixed

RomanDzhabarov commented on source/common/router/req_header_formatter.h:
```diff
@@ -0,0 +1,88 @@
+      Http::HeaderMap& headers, const Http::AccessLog::RequestInfo& requestInfo,
+      const std::list<std::pair<Http::LowerCaseString, std::string>>& requestHeadersToAdd) const;
+
+  std::unordered_map<Http::LowerCaseString, HeaderFormatterPtr, Http::LowerCaseStringHasher>&
```
const, you do not want the caller to modify internal map

khgandhi replied:
ok. i will address this.

khgandhi commented:
@RomanDzhabarov @mattklein123 I have rebased the PR and also addressed all the feedback. Please review it and provide feedback.

@RomanDzhabarov As mentioned in the comments, i was not able to get the code working with changing HeaderFormatter into a unique_ptr. Please review.

RomanDzhabarov commented:
thanks! @khgandhi I'll see today how to fix those unique_ptr errors locally.

RomanDzhabarov commented:
@khgandhi i do not think i'm able to push to your branch. Here is essentially what you need to do:
https://github.com/lyft/envoy/commit/4dd4ad307a235de2fc4615e96de3f234271484c1

Also, tests need to be fixed. In summary, you copied RequestHeaderParser instance which contains unordered map with unique ptr :-)

khgandhi commented:
@RomanDzhabarov Thanks for taking a look at it. However, with those changes,  i see the following error:

source/common/router/config_impl.cc:90:31: error: moving a temporary object prevents copy elision [-Werror,-Wpessimizing-move]
      request_headers_parser_(std::move(RequestHeaderParser::parse(route))),
                              ^
source/common/router/config_impl.cc:90:31: note: remove std::move call here
      request_headers_parser_(std::move(RequestHeaderParser::parse(route))),
                              ^~~~~~~~~~                                 ~
source/common/router/config_impl.cc:423:31: error: moving a temporary object prevents copy elision [-Werror,-Wpessimizing-move]
      request_headers_parser_(std::move(RequestHeaderParser::parse(virtual_host))) {
                              ^
source/common/router/config_impl.cc:423:31: note: remove std::move call here
      request_headers_parser_(std::move(RequestHeaderParser::parse(virtual_host))) {
                              ^~~~~~~~~~                                        ~
source/common/router/config_impl.cc:659:29: error: moving a temporary object prevents copy elision [-Werror,-Wpessimizing-move]
  request_headers_parser_ = std::move(RequestHeaderParser::parse(config));
                            ^
source/common/router/config_impl.cc:659:29: note: remove std::move call here
  request_headers_parser_ = std::move(RequestHeaderParser::parse(config));
                            ^~~~~~~~~~                                  ~

khgandhi commented:
@RomanDzhabarov removing the std::move from the iintializer list work. I will fix the test cases and rebase and submit a new patch.

RomanDzhabarov commented:
awesome, thanks @khgandhi we are very close to finish this review.

RomanDzhabarov commented on source/common/router/req_header_formatter.cc:
```diff
@@ -0,0 +1,93 @@
+
+RequestHeaderParser RequestHeaderParser::parse(const Json::Object& config) {
+  RequestHeaderParser request_header_parser;
+  std::unordered_map<Http::LowerCaseString, HeaderFormatterPtr, Http::LowerCaseStringHasher>
```
no used.

RomanDzhabarov commented on source/common/router/req_header_formatter.h:
```diff
@@ -0,0 +1,89 @@
+typedef std::shared_ptr<HeaderFormatter> HeaderFormatterPtr;
+
+/**
+ * a formatter that expands the request header variable to a value based on info in RequestInfo.
```
nit: starts with capital letter. .... "based on RequestInfo".

RomanDzhabarov commented on source/common/router/req_header_formatter.h:
```diff
@@ -0,0 +1,89 @@
+
+private:
+  /**
+   * building a map of request header formatters.
```
nit: can you check everywhere that sentence starts with capital letter.

RomanDzhabarov commented on source/common/router/req_header_formatter.h:
```diff
@@ -0,0 +1,89 @@
+      const std::list<std::pair<Http::LowerCaseString, std::string>>& requestHeadersToAdd) const;
+
+  const std::unordered_map<Http::LowerCaseString, HeaderFormatterPtr, Http::LowerCaseStringHasher>&
+  headerFormatterMap() {
```
i do not see this is called :-(

RomanDzhabarov commented on source/common/router/req_header_formatter.cc:
```diff
@@ -0,0 +1,93 @@
+    variable_name = format.substr(1, last_occ_pos - 1);
+
+  } else {
+    HeaderFormatterPtr plain_header_formatter_shared_ptr(new PlainHeaderFormatter(format));
```
change all these names, remove shared_ptr part completely.

RomanDzhabarov commented on source/common/router/req_header_formatter.cc:
```diff
@@ -0,0 +1,93 @@
+    HeaderFormatterPtr plain_header_formatter_shared_ptr(new PlainHeaderFormatter(format));
+    return plain_header_formatter_shared_ptr;
+  }
+  HeaderFormatterPtr request_header_formatter_shared_ptr(new RequestHeaderFormatter(variable_name));
```
nit: fix name here as well

khgandhi commented:
@RomanDzhabarov I have merged the PR to the master and resolved all the conflicts.. Also, i have cleaned up the variable names and removed unused function. Please review.

RomanDzhabarov commented:
Thanks, looking

khgandhi commented on source/common/router/req_header_formatter.cc:
```diff
@@ -0,0 +1,120 @@
+  return request_header_parser;
+}
+
+RequestHeaderParserPtr
```
@RomanDzhabarov i tried doing that and ran into an issue with the compiler. it would be great if you could provide some guidance on that:

the three methods take in different proto generated objects (Route, VirtualHost, and RouteConfiguration) and each of them have a method called request_headers_to_add(). This method returns an HeaderValueOption which is not iterable.

**RequestHeaderParserPtr RequestHeaderParser::parseRoute(const envoy::api::v2::Route& route) {**

**RequestHeaderParserPtr RequestHeaderParser::parseVirtualHost(const envoy::api::v2::VirtualHost& virtualHost) {**


**RequestHeaderParserPtr RequestHeaderParser::parseRouteConfiguration(
    const envoy::api::v2::RouteConfiguration& routeConfig) {**


i tried doing the following but kept getting a compilation error:
**RequestHeaderParserPtr RequestHeaderParser::parse(const envoy::api::v2::HeaderValueOption& headers) {
...**

and have the following invocation in config_impl.cc.....
**RequestHeaderParser::parse(routeConfig.request_headers_to_add())..**

 **RequestHeaderParser::parse(route.route().request_headers_to_add())..**

**RequestHeaderParser::parse( virtualHost.request_headers_to_add())...**

the error that i see is

`source/common/router/req_header_formatter.cc:40:42: error: invalid range expression of type 'const envoy::api::v2::HeaderValueOption'; no viable 'begin' function available
    for (const auto& header_value_option : headers) {
....
`

RomanDzhabarov replied:
i believe request_headers_to_add() should return some sort of iterable, can you check that? (not the HeaderValueOption). i think the problem is in the wrong type you are using.

khgandhi replied:
thanks for the pointer.. @RomanDzhabarov .. it is now working.

khgandhi commented on test/common/router/req_header_formatter_test.cc:
```diff
@@ -0,0 +1,104 @@
+using testing::ReturnRef;
+using testing::_;
+
+static envoy::api::v2::Route parseRouteFromJson(const std::string& json_string) {
```
yes.. req_header_formatter_test have ut's for those cases as well.

khgandhi commented on source/common/router/req_header_formatter.cc:
```diff
@@ -0,0 +1,120 @@
+  return request_header_parser;
+}
+
+RequestHeaderParserPtr RequestHeaderParser::parseRouteConfiguration(
```
will address them

khgandhi commented on source/common/router/req_header_formatter.cc:
```diff
@@ -0,0 +1,120 @@
+RequestHeaderParserPtr
+RequestHeaderParser::parseVirtualHost(const envoy::api::v2::VirtualHost& virtualHost) {
+  RequestHeaderParserPtr request_header_parser(new RequestHeaderParser());
+  std::unordered_map<Http::LowerCaseString, HeaderFormatterPtr, Http::LowerCaseStringHasher>
```
will address them

khgandhi commented:
@RomanDzhabarov updated the PR. Please review.

RomanDzhabarov commented on docs/configuration/http_conn_man/route_config/route.rst:
```diff
@@ -400,6 +400,22 @@ specified in the following form:
+The Original Client IP and Protocol are already added by envoy as an  
+:ref:`X-Forwarded-For <config_http_conn_man_headers_x-forwarded-for>` 
+and :ref:`X-Forwarded-Proto <config_http_conn_man_headers_x-forwarded-proto>` headers. 
+The above feature lets you customize the header name to be added to the request(X-Client-IP 
```
to the request headers.

RomanDzhabarov commented on source/common/router/config_impl.h:
```diff
@@ -86,6 +88,8 @@ class VirtualHostImpl : public VirtualHost {
   const std::string& name() const override { return name_; }
   const RateLimitPolicy& rateLimitPolicy() const override { return rate_limit_policy_; }
 
+  RequestHeaderParser& requestHeaderParser() const { return *request_headers_parser_; };
```
I find it strange that the method is declared as `const` but returns a reference to internal data.
I'd remove `const` in here.

RomanDzhabarov replied:
this is not addressed.

khgandhi replied:
if i remove the const, i see the following error:

source/common/router/config_impl.cc:159:3: error: member function 'requestHeaderParser' not viable: 'this' argument has type 'const Envoy::Router::VirtualHostImpl', but function is not marked const
  vhost_.requestHeaderParser().evaluateRequestHeaders(headers, requestInfo,
  ^~~~~~

RomanDzhabarov replied:
ah, my bad... if this does not work could you try `return const RequestHeaderParser&`.
In general, you should not expose a reference to class internals in `const` method.

RomanDzhabarov commented on source/common/router/req_header_formatter.h:
```diff
@@ -0,0 +1,97 @@
+};
+
+/**
+ * Returns back the same static header value.
```
nit: can you use the same style of comment as above. A formatter ...

RomanDzhabarov commented on source/common/router/req_header_formatter.h:
```diff
@@ -0,0 +1,97 @@
+  static RequestHeaderParserPtr
+  parseRouteConfiguration(const envoy::api::v2::RouteConfiguration& routeConfig);
+
+  static RequestHeaderParserPtr
```
this is private.

RomanDzhabarov commented on test/common/router/req_header_formatter_test.cc:
```diff
@@ -0,0 +1,104 @@
+  std::string s1 = "127.0.0.1";
+  ON_CALL(requestInfo, getDownstreamAddress()).WillByDefault(ReturnRef(s1));
+  std::string variable = "INVALID_VARIABLE";
+  ASSERT_THROW(RequestHeaderFormatter requestHeaderFormatter(variable), EnvoyException);
```
can you assert throw with error message? (see corresponding macro)

RomanDzhabarov replied:
i'm sorry but i still do not see the test case where `last_occ_pos <= 1` is exercised.

khgandhi replied:
@RomanDzhabarov ok.. i will a ut for that as well.

RomanDzhabarov commented:
cc: @mattklein123 if you want to do another pass. I have only a few nits left in here.

khgandhi commented:
@RomanDzhabarov Thanks for the LGTM. i have addressed all the feedback.. @mattklein123 please review the PR.

mattklein123 commented on docs/configuration/http_conn_man/route_config/route.rst:
```diff
@@ -400,6 +400,22 @@ specified in the following form:
     {"key": "header2", "value": "value2"}
   ]
 
+Envoy also supports adding the Original Client IP and Protocol to the request headers that matches a 
```
It would be nice if you could rework these docs a bit to make it a bit more like: https://lyft.github.io/envoy/docs/configuration/http_conn_man/access_log.html#format-rules

We will add new rules in the future and that format will make it easier to do so.

mattklein123 commented on source/common/http/access_log/request_info_impl.h:
```diff
@@ -46,6 +46,11 @@ struct RequestInfoImpl : public RequestInfo {
 
   void healthCheck(bool is_hc) override { hc_request_ = is_hc; }
 
+  /**
```
nit: remove comment (we don't duplicate comments for overrides).

mattklein123 commented on source/common/http/websocket/ws_handler_impl.cc:
```diff
@@ -13,12 +13,12 @@ namespace Envoy {
 
-WsHandlerImpl::WsHandlerImpl(HeaderMap& request_headers, const Router::RouteEntry& route_entry,
-                             WsHandlerCallbacks& callbacks,
+WsHandlerImpl::WsHandlerImpl(HeaderMap& request_headers, AccessLog::RequestInfo& request_info,
```
`const AccessLog::RequestInfo&`

mattklein123 commented on source/common/router/config_impl.h:
```diff
@@ -15,7 +15,9 @@
 #include "envoy/runtime/runtime.h"
 #include "envoy/upstream/cluster_manager.h"
 
+#include "common/http/access_log/access_log_formatter.h"
```
Is this header needed? What is it needed for? If not needed please remember to remove from BUILD file also.

khgandhi replied:
This is one is not required.

mattklein123 commented on source/common/router/req_header_formatter.h:
```diff
@@ -0,0 +1,97 @@
+#include <unordered_map>
+
+#include "common/common/logger.h"
+#include "common/http/access_log/access_log_formatter.h"
```
Is this needed? If so what for?

khgandhi replied:
we need the #include "envoy/http/access_log.h" header for the RequestInfo class. I can replace the access_log_formatter.h with #include "envoy/http/access_log.h"

khgandhi replied:
i guess it is required for Envoy::Http::AccessLog::AccessLogFormatUtils::protocolToString(..) method in req_header_formatter.cc

mattklein123 commented on source/common/router/req_header_formatter.h:
```diff
@@ -0,0 +1,97 @@
+public:
+  virtual ~RequestHeaderParser() {}
+
+  static RequestHeaderParserPtr parseRoute(const envoy::api::v2::Route& route);
```
I would delete parseRoute, parseVirtualHost, parseRouteConfiguration, move parse() to public, and then just call with the needed proto member at callsite. I don't think the extra indirection is needed.

mattklein123 commented on source/common/router/req_header_formatter.h:
```diff
@@ -0,0 +1,97 @@
+      const std::list<std::pair<Http::LowerCaseString, std::string>>& requestHeadersToAdd) const;
+
+private:
+  /**
```
nit: del obvious comment

mattklein123 commented on source/common/router/req_header_formatter.cc:
```diff
@@ -0,0 +1,98 @@
+
+HeaderFormatterPtr RequestHeaderParser::parseInternal(const std::string& format) {
+  if (format.find("%") == 0) {
+    size_t last_occ_pos = format.rfind("%");
```
const size_t

mattklein123 commented on source/common/router/req_header_formatter.cc:
```diff
@@ -0,0 +1,98 @@
+    }
+    const std::string variable_name = format.substr(1, last_occ_pos - 1);
+    HeaderFormatterPtr request_header_formatter_ptr(new RequestHeaderFormatter(variable_name));
+    return request_header_formatter_ptr;
```
nit: 1 liner: `return HeaderFormatterPtr{new ...}` same below

mattklein123 commented on test/common/router/req_header_formatter_test.cc:
```diff
@@ -0,0 +1,135 @@
+
+namespace Envoy {
+namespace Router {
+using testing::NiceMock;
```
nit: please move using outside of the namespace decls.

mattklein123 commented on test/common/router/config_impl_test.cc:
```diff
@@ -2031,6 +2036,228 @@ TEST(RoutePropertyTest, excludeVHRateLimits) {
+  NiceMock<Runtime::MockLoader> runtime;
+  NiceMock<Upstream::MockClusterManager> cm;
+  NiceMock<Envoy::Http::AccessLog::MockRequestInfo> requestInfo;
+  ASSERT_THROW(ConfigImpl config(parseRouteConfigurationFromJson(json), runtime, cm, true),
```
s/ASSERT_THROW/EXPECT_THROW (elsewhere also)

mattklein123 commented on test/common/router/config_impl_test.cc:
```diff
@@ -2031,6 +2036,228 @@ TEST(RoutePropertyTest, excludeVHRateLimits) {
+  }
+}
+
+bool CheckFormatMessage(const std::string errorMessage) {
```
nit: `static bool CheckFormatMessage`, use snake case for param (please check everywhere else for camelCase that should be snake_case)

mattklein123 commented on test/common/router/req_header_formatter_test.cc:
```diff
@@ -0,0 +1,135 @@
+  return route;
+}
+
+bool CheckFormatMessage(const std::string errorMessage, const std::string expectedMessage) {
```
same comments about style as other file.

khgandhi commented:
@mattklein123 addressed all your feedback. Please review

RomanDzhabarov commented on docs/configuration/http_conn_man/route_config/route.rst:
```diff
@@ -400,6 +400,24 @@ specified in the following form:
+.. code-block:: json
+
+  [
+   {"key": "X-Client-IP", "value":"%CLIENT_IP%"}
```
can you also add integration test to test this? (need to change integration test envoy config)

khgandhi replied:
can you give pointers on which integration test do we need change ?

RomanDzhabarov replied:
I'm thinking maybe here: https://github.com/lyft/envoy/blob/70c5bc94adaed8d94f47cd282f1ae0a11bf08743/test/tools/router_check/test/config/TestRoutes.json#L2

khgandhi replied:
got it.. @RomanDzhabarov thanks for the pointer.. I will add some tests there.

RomanDzhabarov commented on test/common/router/req_header_formatter_test.cc:
```diff
@@ -0,0 +1,138 @@
+  req_header_parser->evaluateRequestHeaders(headerMap, requestInfo, requestHeadersToAdd);
+  EXPECT_TRUE(headerMap.has("static-header"));
+}
+} // namespace Router
```
new line between } and } // namespace Router.

RomanDzhabarov commented on test/common/router/req_header_formatter_test.cc:
```diff
@@ -0,0 +1,138 @@
+
+static bool CheckFormatMessage(const std::string error_message,
+                               const std::string expected_message) {
+  std::size_t found = error_message.find(expected_message);
```
const :)

khgandhi commented:
@RomanDzhabarov In config_impl.h, i had to mark the requestHeaderParser() as a const because the VirtualHostImpl object is a const. If i remove the const, i see the following error:

source/common/router/config_impl.cc:159:3: error: member function 'requestHeaderParser' not viable: 'this' argument has type 'const Envoy::Router::VirtualHostImpl', but function is not marked const
  vhost_.requestHeaderParser().evaluateRequestHeaders(headers, requestInfo,
  ^~~~~~

mattklein123 commented:
@khgandhi per @RomanDzhabarov return const request header parser. I will look again when @RomanDzhabarov comments are addressed and PR is passing tests. Thank you.

mattklein123 commented:
@khgandhi I see that you rebased. Nothing to do about it now, but in the future, please, please, please don't rebase once you open the PR. It's much harder to review the changes. Thank you!

khgandhi commented:
@mattklein123 ok.. i will  keep that in mind.. @RomanDzhabarov @mattklein123 can you please review the changes.. i have added integration test case and addressed other comments as well..

mattklein123 commented on docs/configuration/http_conn_man/route_config/route.rst:
```diff
@@ -400,6 +400,24 @@ specified in the following form:
+Envoy supports adding static and dynamic values to the request headers. The dynamic values supported are:
+
+%CLIENT_IP%
+   The Original Client IP which is already added by envoy as an  
```
nit: "original client IP"

mattklein123 commented on docs/configuration/http_conn_man/route_config/route.rst:
```diff
@@ -400,6 +400,24 @@ specified in the following form:
+   :ref:`X-Forwarded-For <config_http_conn_man_headers_x-forwarded-for>` request header. 
+
+%PROTOCOL%
+    The Original Protocol which is already added by envoy as an 
```
nit: "original protocol"

mattklein123 commented on include/envoy/router/router.h:
```diff
@@ -187,7 +188,8 @@ class RouteEntry {
    */
-  virtual void finalizeRequestHeaders(Http::HeaderMap& headers) const PURE;
+  virtual void finalizeRequestHeaders(Http::HeaderMap& headers,
+                                      const Http::AccessLog::RequestInfo& requestInfo) const PURE;
```
request_info. Please update doc comments.

mattklein123 commented on source/common/router/config_impl.cc:
```diff
@@ -332,8 +330,9 @@ PrefixRouteEntryImpl::PrefixRouteEntryImpl(const VirtualHostImpl& vhost,
-void PrefixRouteEntryImpl::finalizeRequestHeaders(Http::HeaderMap& headers) const {
-  RouteEntryImplBase::finalizeRequestHeaders(headers);
+void PrefixRouteEntryImpl::finalizeRequestHeaders(
+    Http::HeaderMap& headers, const Http::AccessLog::RequestInfo& requestInfo) const {
```
request_info

mattklein123 commented on source/common/router/config_impl.cc:
```diff
@@ -351,8 +350,9 @@ PathRouteEntryImpl::PathRouteEntryImpl(const VirtualHostImpl& vhost,
-void PathRouteEntryImpl::finalizeRequestHeaders(Http::HeaderMap& headers) const {
-  RouteEntryImplBase::finalizeRequestHeaders(headers);
+void PathRouteEntryImpl::finalizeRequestHeaders(
+    Http::HeaderMap& headers, const Http::AccessLog::RequestInfo& requestInfo) const {
```
request_info

mattklein123 commented on source/common/router/config_impl.h:
```diff
@@ -194,7 +198,9 @@ class RouteEntryImplBase : public RouteEntry,
   const std::string& clusterName() const override;
-  void finalizeRequestHeaders(Http::HeaderMap& headers) const override;
+  void finalizeRequestHeaders(Http::HeaderMap& headers,
+                              const Http::AccessLog::RequestInfo& requestInfo) const override;
```
request_info (please do global find and replace in case I missed any)

mattklein123 commented on source/common/router/config_impl.h:
```diff
@@ -242,8 +249,9 @@ class RouteEntryImplBase : public RouteEntry,
-    void finalizeRequestHeaders(Http::HeaderMap& headers) const override {
-      return parent_->finalizeRequestHeaders(headers);
+    void finalizeRequestHeaders(Http::HeaderMap& headers,
+                                const Http::AccessLog::RequestInfo& requestInfo) const override {
```
request_info (I'm going to stop commenting now. do global find/replace)

mattklein123 commented on source/common/router/req_header_formatter.cc:
```diff
@@ -0,0 +1,81 @@
+    const size_t last_occ_pos = format.rfind("%");
+    if (last_occ_pos == std::string::npos || last_occ_pos <= 1) {
+      throw EnvoyException(fmt::format(
+          "Incorrect configuration: {}. Expected the variable to be of format %<variable_name>%",
```
nit: Incorrect header variable configuration

mattklein123 commented on source/common/router/req_header_formatter.cc:
```diff
@@ -0,0 +1,81 @@
+    const Protobuf::RepeatedPtrField<envoy::api::v2::HeaderValueOption>& headers) {
+  RequestHeaderParserPtr request_header_parser(new RequestHeaderParser());
+  for (const auto& header_value_option : headers) {
+    ENVOY_LOG(debug, "adding key {} to header formatter map", header_value_option.header().key());
```
debug message not very useful. I would remove.

mattklein123 commented on source/common/router/req_header_formatter.cc:
```diff
@@ -0,0 +1,81 @@
+    Http::HeaderMap& headers, const Http::AccessLog::RequestInfo& requestInfo,
+    const std::list<std::pair<Http::LowerCaseString, std::string>>& requestHeadersToAdd) const {
+  for (const auto& to_add : requestHeadersToAdd) {
+    ENVOY_LOG(debug, "request headers key {}", to_add.first.get());
```
debug message not very useful. I would remove.

mattklein123 commented on source/common/router/req_header_formatter.cc:
```diff
@@ -0,0 +1,81 @@
+
+void RequestHeaderParser::evaluateRequestHeaders(
+    Http::HeaderMap& headers, const Http::AccessLog::RequestInfo& requestInfo,
+    const std::list<std::pair<Http::LowerCaseString, std::string>>& requestHeadersToAdd) const {
```
request_headers_to_add (please go through your entire diff and look for snake_case issues. Thanks)

mattklein123 commented on source/common/router/req_header_formatter.cc:
```diff
@@ -0,0 +1,81 @@
+void RequestHeaderParser::evaluateRequestHeaders(
+    Http::HeaderMap& headers, const Http::AccessLog::RequestInfo& requestInfo,
+    const std::list<std::pair<Http::LowerCaseString, std::string>>& requestHeadersToAdd) const {
+  for (const auto& to_add : requestHeadersToAdd) {
```
snake_case

mattklein123 commented on source/common/router/req_header_formatter.h:
```diff
@@ -0,0 +1,87 @@
+      const std::list<std::pair<Http::LowerCaseString, std::string>>& requestHeadersToAdd) const;
+
+private:
+  std::unordered_map<Http::LowerCaseString, HeaderFormatterPtr, Http::LowerCaseStringHasher>
```
Sorry I'm coming to this very late, but I don't understand why we are using a map here and having to do any kind of lookup. You do a pre-parse. You should just end up with a list of things that you need to add. The map lookup at runtime I think is not needed?

khgandhi replied:
@mattklein123 .. The map is there to store the pre-parsed formatter for each of the field in the request_headers_to_add configuration.

During evaluation phase (evaluateRequestHeaders method), the code just goes through the request_headers_to_add list and evaluate the fields that are in the pre-parsed.

mattklein123 replied:
The way to implement this is to pre-parse into a list of actions, including headers that just end up getting statically added. Then during runtime, you just run through the list and fill them in. No map is needed. This will be more efficient.

khgandhi replied:
@mattklein123 yes.. i was able to get rid of the map and replace it with a list of formatters (dynamic and static) and just run through them. I will update the PR with that change and rest of the feedbacks tonight.. Thanks

mattklein123 commented on test/common/router/req_header_formatter_test.cc:
```diff
@@ -0,0 +1,139 @@
+#include "gtest/gtest.h"
+
+namespace Envoy {
+using testing::NiceMock;
```
Move usings outside of namespace

mattklein123 commented on test/tools/router_check/router.cc:
```diff
@@ -60,6 +60,7 @@ RouterCheckTool::RouterCheckTool(std::unique_ptr<NiceMock<Runtime::MockLoader>>
     : runtime_(std::move(runtime)), cm_(std::move(cm)), config_(std::move(config)) {}
 
 bool RouterCheckTool::compareEntriesInJson(const std::string& expected_route_json) {
+
```
del

mattklein123 commented on test/tools/router_check/test/config/TestRoutes.golden.json:
```diff
@@ -288,5 +288,21 @@
     }
-  }
+  },
+ {
```
indent messed up

mattklein123 commented on test/tools/router_check/test/config/TestRoutes.json:
```diff
@@ -87,11 +87,20 @@
           "cluster": "instant-server",
           "case_sensitive": true
         },
+	{
```
indent messed up

khgandhi commented:
@mattklein123 incorporated all the feedback.. please take a look.

mattklein123 commented on source/common/router/req_header_formatter.cc:
```diff
@@ -0,0 +1,73 @@
+void RequestHeaderParser::evaluateRequestHeaders(
+    Http::HeaderMap& headers, const Http::AccessLog::RequestInfo& request_info) const {
+  for (const auto& formatter : header_formatters_) {
+    const std::string formatted_header_value = formatter.second->format(request_info);
```
nit: local var not needed, just do inline.

mattklein123 commented on source/common/router/router.cc:
```diff
@@ -228,8 +228,7 @@ Http::FilterHeadersStatus Filter::decodeHeaders(Http::HeaderMap& headers, bool e
     timeout_response_code_ = Http::Code::NoContent;
     headers.removeEnvoyUpstreamRequestTimeoutAltResponse();
   }
-
```
revert newline del

mattklein123 commented on test/common/router/config_impl_test.cc:
```diff
@@ -2031,6 +2036,228 @@ TEST(RoutePropertyTest, excludeVHRateLimits) {
+TEST(CustomRequestHeadersTest, AddNewHeader) {
+  const std::string json = R"EOF(
+	{
+  "virtual_hosts": [
```
indent messed up

mattklein123 commented on test/common/router/config_impl_test.cc:
```diff
@@ -2031,6 +2036,228 @@ TEST(RoutePropertyTest, excludeVHRateLimits) {
+  }
+}
+
+static bool CheckFormatMessage(const std::string errorMessage) {
```
Already commented on this: `checkFormatMessage `

mattklein123 commented on test/common/router/config_impl_test.cc:
```diff
@@ -2031,6 +2036,228 @@ TEST(RoutePropertyTest, excludeVHRateLimits) {
+TEST(CustomRequestHeadersTest, CustomHeaderWrongFormat) {
+  std::string json = R"EOF(
+	{
+  "virtual_hosts": [
```
indent messed up

mattklein123 commented on test/common/router/req_header_formatter_test.cc:
```diff
@@ -0,0 +1,135 @@
+  return route;
+}
+
+static bool CheckFormatMessage(const std::string error_message,
```
already commented on this. Functions start with lower case.

mattklein123 commented on test/common/router/req_header_formatter_test.cc:
```diff
@@ -0,0 +1,135 @@
+}
+
+TEST(RequestHeaderFormatterTest, WrongFormatOnVariable) {
+  std::string json = R"EOF({
```
please 2 space indent the JSON like everywhere else. Same below.

mattklein123 commented on test/tools/router_check/test/config/TestRoutes.json:
```diff
@@ -87,11 +87,20 @@
+          "prefix": "/customheaders",
+          "cluster": "ats",
+          "host_rewrite": "new_host",
+	  "request_headers_to_add":[
```
indent still messed up............

khgandhi commented:
@mattklein123 Thanks for the feedback.. I will look at the whole PR and clean up all the formatting issues to make sure it is at par with the current code's quality.

RomanDzhabarov commented on source/common/router/req_header_formatter.cc:
```diff
@@ -0,0 +1,73 @@
+#include "common/router/req_header_formatter.h"
+
+#include <cstdint>
+#include <iostream>
```
iostream?

RomanDzhabarov commented on source/common/router/req_header_formatter.cc:
```diff
@@ -0,0 +1,73 @@
+#include <cstdint>
+#include <iostream>
+#include <string>
+#include <unordered_map>
```
probably not used, can you inspect all the includes here.

RomanDzhabarov commented on source/common/router/req_header_formatter.cc:
```diff
@@ -0,0 +1,73 @@
+  if (format.find("%") == 0) {
+    const size_t last_occ_pos = format.rfind("%");
+    if (last_occ_pos == std::string::npos || last_occ_pos <= 1) {
+      throw EnvoyException(fmt::format("Incorrect header configuration: {}. Expected the variable "
```
I'd also add.. `expected "A" but given "B"` to make this message more descriptive.

RomanDzhabarov commented on source/common/router/req_header_formatter.h:
```diff
@@ -0,0 +1,85 @@
+#pragma once
+
+#include <cstdint>
+#include <iostream>
```
please clean includes

RomanDzhabarov commented on source/common/router/req_header_formatter.h:
```diff
@@ -0,0 +1,85 @@
+typedef std::unique_ptr<RequestHeaderParser> RequestHeaderParserPtr;
+
+/**
+ * This class will hold the parsing logic required during configuration build and
```
nit: s/will hold/holds

RomanDzhabarov commented on test/common/router/req_header_formatter_test.cc:
```diff
@@ -0,0 +1,135 @@
+
+TEST(RequestHeaderFormatterTest, TestFormatWithClientIpVariable) {
+  NiceMock<Envoy::Http::AccessLog::MockRequestInfo> request_info;
+  std::string s1 = "127.0.0.1";
```
very minor: can you go through the file and fix `const` where applicable.

RomanDzhabarov commented on test/common/router/req_header_formatter_test.cc:
```diff
@@ -0,0 +1,135 @@
+  ON_CALL(request_info, getDownstreamAddress()).WillByDefault(ReturnRef(s1));
+  std::string variable = "INVALID_VARIABLE";
+  ASSERT_THROW(RequestHeaderFormatter requestHeaderFormatter(variable), EnvoyException);
+  try {
```
sorry, i should have been more clear here, i meant to use EXPECT_THROW_WITH_MESSAGE
https://github.com/lyft/envoy/search?utf8=%E2%9C%93&q=EXPECT_THROW_WITH_MESSAGE&type=

RomanDzhabarov commented on test/common/router/req_header_formatter_test.cc:
```diff
@@ -0,0 +1,135 @@
+            }
+          ]
+        })EOF";
+  ASSERT_THROW(Envoy::Router::RequestHeaderParser::parse(
```
no assert, expect

RomanDzhabarov commented on test/common/router/req_header_formatter_test.cc:
```diff
@@ -0,0 +1,135 @@
+  try {
+    Envoy::Router::RequestHeaderParser::parse(
+        parseRouteFromJson(json).route().request_headers_to_add());
+  } catch (EnvoyException& ex) {
```
same https://github.com/lyft/envoy/search?utf8=%E2%9C%93&q=EXPECT_THROW_WITH_MESSAGE&type=

mattklein123 commented:
Closing for now. Let's reopen when we are ready to get this in.
